field/src/field.rs

```diff
@@ -551,12 +551,138 @@ pub trait Algebra<F>:
 // Every ring is an algebra over itself.
 impl<R: PrimeCharacteristicRing> Algebra<R> for R {}
 
+/// A collection of methods designed to help hash field elements.
+///
+/// Most fields will want to reimplement many/all of these methods as the default implementations
+/// are slow and involve converting to/from byte representations.
+pub trait RawDataSerializable: Sized {
+    /// The number of bytes which this field element occupies in memory.
+    /// Must be equal to the length of self.into_bytes().
+    const NUM_BYTES: usize;
+
+    /// Convert a field element into a collection of bytes.
+    #[must_use]
+    fn into_bytes(self) -> impl IntoIterator<Item = u8>;
+
+    /// Convert an iterator of field elements into an iterator of bytes.
+    #[must_use]
+    fn into_byte_stream(input: impl IntoIterator<Item = Self>) -> impl IntoIterator<Item = u8> {
+        input.into_iter().flat_map(|elem| elem.into_bytes())
+    }
+
+    /// Convert an iterator of field elements into an iterator of u32s.
+    ///
+    /// If `NUM_BYTES` does not divide `4`, multiple `F`s may be packed together to make a single `u32`. Furthermore,
+    /// if `NUM_BYTES * input.len()` does not divide `4`, the final `u32` will involve padding bytes which are set to `0`.
+    #[must_use]
+    fn into_u32_stream(input: impl IntoIterator<Item = Self>) -> impl IntoIterator<Item = u32> {
+        let mut bytes = Self::into_byte_stream(input).into_iter().peekable();
+
+        iter::from_fn(move || {
+            bytes.peek().is_some().then(|| {
+                let u32_bytes = array::from_fn(|_| bytes.next().unwrap_or_default());
+                u32::from_le_bytes(u32_bytes)
+            })
+        })
```

Comment:
In a glorious future this will be the much more idiomatic
```suggestion
        let bytes = Self::into_byte_stream(input).into_iter();

        bytes
            .chain(iter::repeat(0))
            .array_chunks::<4>()
            .map(u32::from_le_bytes)
```
Alas, not today.

You could consider copying `.array_chunks` from nightly though, part of it is already there in `p3_utils`...